src/chk/chk_leader.c

```diff
@@ -0,0 +1,1197 @@
+
+log:
+	if (rc != 0) {
+		D_ERROR("Check leader (gen "DF_X64") failed to handle %s report from rank %u with "
+			"seq "DF_X64", class %u, action %u, result %d: "DF_RC"\n", cbk->cb_gen,
+			local ? "local" : "remote", rank, *seq, cla, act, result, DP_RC(rc));
+		goto out;
+	}
+
+	D_CDEBUG(act == CHK__CHECK_INCONSIST_ACTION__CIA_INTERACT, DLOG_INFO, DLOG_DBG,
+		 "Check leader (gen "DF_X64") handle %s report from rank %u with seq "
+		 DF_X64" class %u, action %u, result %d: "DF_RC"\n",
+		 cbk->cb_gen, local ? "local" : "remote", rank, *seq, cla, act, result, DP_RC(rc));
+
+	if (!local)
+		goto out;
+
+	D_ASSERT(cpr->cpr_busy == 1);
+
+	ABT_mutex_lock(cpr->cpr_mutex);
+	if (cpr->cpr_action != CHK__CHECK_INCONSIST_ACTION__CIA_INTERACT) {
+		ABT_mutex_unlock(cpr->cpr_mutex);
+		goto handle;
+	}
+
+	ABT_cond_wait(cpr->cpr_cond, cpr->cpr_mutex);
+	ABT_mutex_unlock(cpr->cpr_mutex);
+	if (!chk_leader->ci_sched_running || cpr->cpr_exiting)
+		goto out;
+
+	D_ASSERT(cpr->cpr_action != CHK__CHECK_INCONSIST_ACTION__CIA_INTERACT);
+
+handle:
+
+	/* TBD: already got the interaction feedback, handle it and log the result. */
+
+out:
+	if (cpr != NULL)
+		chk_pending_destroy(cpr);
+
+	return rc;
+}
+
+int
+chk_leader_notify(uint64_t gen, d_rank_t rank, uint32_t phase, uint32_t status)
+{
+	struct chk_property	*prop = &chk_leader->ci_prop;
+	struct chk_bookmark	*cbk = &chk_leader->ci_bk;
+	struct chk_rank_bundle	 rbund = { 0 };
+	d_iov_t			 kiov;
+	d_iov_t			 riov;
+	int			 rc = 0;
+
+	if (cbk->cb_gen != gen)
+		D_GOTO(out, rc = -DER_NOTAPPLICABLE);
+
+	if (cbk->cb_ins_status != CHK__CHECK_INST_STATUS__CIS_RUNNING)
+		D_GOTO(out, rc = -DER_NOTAPPLICABLE);
+
+	switch (status) {
+	case CHK__CHECK_INST_STATUS__CIS_INIT:
+	case CHK__CHECK_INST_STATUS__CIS_STOPPED:
+	case CHK__CHECK_INST_STATUS__CIS_PAUSED:
+	case CHK__CHECK_INST_STATUS__CIS_IMPLICATED:
+		/* Directly ignore above. */
+		D_GOTO(out, rc = 0);
+	case CHK__CHECK_INST_STATUS__CIS_RUNNING:
+		if (unlikely(phase < cbk->cb_phase))
+			D_GOTO(log, rc = -DER_INVAL);
+
+		if (phase == cbk->cb_phase)
+			D_GOTO(out, rc = 0);
+
+		rbund.crb_rank = rank;
+		rbund.crb_phase = phase;
+		rbund.crb_ins = chk_leader;
+
+		d_iov_set(&riov, &rbund, sizeof(rbund));
+		d_iov_set(&kiov, &rank, sizeof(rank));
+		rc = dbtree_upsert(chk_leader->ci_rank_hdl, BTR_PROBE_EQ, DAOS_INTENT_UPDATE,
+				   &kiov, &riov, NULL);
+
+		D_GOTO(log, rc);
+	case CHK__CHECK_INST_STATUS__CIS_COMPLETED:
+		/*
+		 * XXX: Currently, we do not support to partial check till the specified phase.
+		 *	Then the completed phase will be either container cleanup or all done.
+		 */
+		if (unlikely(phase != CHK__CHECK_SCAN_PHASE__CSP_CONT_CLEANUP &&
+			     phase != CHK__CHECK_SCAN_PHASE__DSP_DONE))
+			D_GOTO(log, rc = -DER_INVAL);
+
+		d_iov_set(&kiov, &rank, sizeof(rank));
+		rc = dbtree_delete(chk_leader->ci_pending_hdl, BTR_PROBE_EQ, &kiov, NULL);
+
+		D_GOTO(log, rc = (rc == -DER_ENOENT ? 0 : rc));
+	case CHK__CHECK_INST_STATUS__CIS_FAILED:
```

Comment:
(style)  Possible switch case/default not preceeded by break or fallthrough comment